File: backend_archived/app/modules/classificacao_hrb.py
# backend/app/modules/classificacao_hrb.py
"""
Módulo para classificação de solos segundo o sistema HRB (Highway Research Board)
também conhecido como classificação AASHTO (American Association of State Highway 
and Transportation Officials).

Este sistema é amplamente utilizado para classificação de solos para fins de 
pavimentação rodoviária.

Referências:
- AASHTO M 145 - Standard Specification for Classification of Soils and 
  Soil-Aggregate Mixtures for Highway Construction Purposes
- DER/SP - Departamento de Estradas de Rodagem de São Paulo
"""

from typing import Optional
from app.models import ClassificacaoHRBInput, ClassificacaoHRBOutput

EPSILON = 1e-9


def classificar_hrb(dados: ClassificacaoHRBInput) -> ClassificacaoHRBOutput:
    """
    Classifica o solo de acordo com o sistema HRB/AASHTO.
    
    Args:
        dados: Objeto ClassificacaoHRBInput contendo os parâmetros necessários.
        
    Returns:
        Um objeto ClassificacaoHRBOutput com a classificação, descrição e índice de grupo.
    """
    try:
        # Extrai dados de entrada
        p200 = dados.pass_peneira_200  # % passando na peneira #200
        p40 = dados.pass_peneira_40    # % passando na peneira #40
        p10 = dados.pass_peneira_10    # % passando na peneira #10
        ll = dados.ll                   # Limite de Liquidez
        ip = dados.ip                   # Índice de Plasticidade
        
        # Validações
        if not (0 <= p200 <= 100):
            raise ValueError("% passando na #200 deve estar entre 0 e 100.")
        
        # LL e IP são opcionais para alguns grupos, mas necessários para outros
        if ll is not None and ll < 0:
            raise ValueError("LL não pode ser negativo.")
        if ip is not None and ip < 0:
            raise ValueError("IP não pode ser negativo.")
        
        # Determina o grupo principal
        grupo, subgrupo = _determinar_grupo_hrb(p200, p40, p10, ll, ip)
        
        # Calcula o Índice de Grupo (IG)
        ig = _calcular_indice_grupo(p200, ll, ip)
        
        # Monta a classificação completa
        if subgrupo:
            classificacao = f"{grupo}-{subgrupo}"
        else:
            classificacao = grupo
        
        # Adiciona o índice de grupo se aplicável
        if ig > 0:
            classificacao += f" ({ig})"
        
        # Obtém a descrição do grupo
        descricao = _obter_descricao_hrb(grupo, subgrupo)
        
        # Obtém avaliação como subleito
        avaliacao_subleito = _obter_avaliacao_subleito(grupo)
        
        return ClassificacaoHRBOutput(
            classificacao=classificacao,
            grupo_principal=grupo,
            subgrupo=subgrupo,
            indice_grupo=ig,
            descricao=descricao,
            avaliacao_subleito=avaliacao_subleito
        )
        
    except ValueError as ve:
        return ClassificacaoHRBOutput(erro=str(ve))
    except Exception as e:
        print(f"Erro inesperado na classificação HRB: {e}")
        return ClassificacaoHRBOutput(erro=f"Erro interno no servidor: {type(e).__name__}")


def _determinar_grupo_hrb(
    p200: float,
    p40: Optional[float],
    p10: Optional[float],
    ll: Optional[float],
    ip: Optional[float]
) -> tuple[str, Optional[str]]:
    """
    Determina o grupo e subgrupo HRB baseado nos critérios da AASHTO.
    
    Returns:
        Tupla (grupo, subgrupo)
    """
    
    # MATERIAIS GRANULARES (≤ 35% passando na #200)
    if p200 <= 35:
        
        # Grupo A-1
        if p200 <= 15:
            if p40 is not None and p40 <= 50:
                return ("A-1", "a")
            elif p40 is not None and p40 > 50:
                return ("A-1", "b")
            else:
                return ("A-1", None)
        
        # Grupo A-3 (areia fina)
        if p200 <= 10:
            if ll is not None and ll <= 40:
                if ip is None or ip <= 10:
                    # Verifica se tem características de A-3
                    # A-3: mais de 51% passa na #40 e IP ≤ 10
                    if p40 is not None and p40 > 51:
                        return ("A-3", None)
        
        # Grupo A-2 (materiais granulares siltosos ou argilosos)
        if ll is not None and ip is not None:
            if ll <= 40:
                if ip <= 10:
                    return ("A-2", "4")
                else:
                    return ("A-2", "6")
            else:  # LL > 40
                if ip <= 10:
                    return ("A-2", "5")
                else:
                    return ("A-2", "7")
        
        # Se não tem LL e IP, assume A-2-4 como padrão para granulares com finos
        if p200 > 15:
            return ("A-2", "4")
        
        # Padrão para granulares
        return ("A-1", None)
    
    # MATERIAIS SILTO-ARGILOSOS (> 35% passando na #200)
    else:
        if ll is None or ip is None:
            raise ValueError("LL e IP são necessários para classificar solos com mais de 35% de finos.")
        
        # Grupo A-7 (solos argilosos)
        if ll > 40 and ip > 10:
            # A-7-5 se IP ≤ LL - 30
            # A-7-6 se IP > LL - 30
            if ip <= (ll - 30):
                return ("A-7", "5")
            else:
                return ("A-7", "6")
        
        # Grupo A-6 (solos argilosos)
        if ll <= 40 and ip > 10:
            return ("A-6", None)
        
        # Grupo A-5 (solos siltosos)
        if ll > 40 and ip <= 10:
            return ("A-5", None)
        
        # Grupo A-4 (solos siltosos)
        if ll <= 40 and ip <= 10:
            return ("A-4", None)
        
        # Fallback
        return ("A-4", None)


def _calcular_indice_grupo(p200: float, ll: Optional[float], ip: Optional[float]) -> int:
    """
    Calcula o Índice de Grupo (IG) segundo a fórmula AASHTO.
    
    IG = (F - 35)[0.2 + 0.005(LL - 40)] + 0.01(F - 15)(IP - 10)
    
    Onde:
    - F = % passando na peneira #200
    - LL = Limite de Liquidez
    - IP = Índice de Plasticidade
    
    Restrições:
    - IG máximo de cada parcela é 4
    - IG negativo é considerado 0
    - IG é arredondado para o inteiro mais próximo
    """
    
    # IG só é calculado para solos com mais de 35% de finos
    if p200 <= 35:
        return 0
    
    # Se não temos LL ou IP, não podemos calcular IG
    if ll is None or ip is None:
        return 0
    
    # Primeira parcela: (F - 35)[0.2 + 0.005(LL - 40)]
    parcela1 = 0.0
    if p200 > 35 and ll > 40:
        parcela1 = (p200 - 35) * (0.2 + 0.005 * (ll - 40))
        parcela1 = min(parcela1, 4.0)  # Máximo de 4
    
    # Segunda parcela: 0.01(F - 15)(IP - 10)
    parcela2 = 0.0
    if p200 > 15 and ip > 10:
        parcela2 = 0.01 * (p200 - 15) * (ip - 10)
        parcela2 = min(parcela2, 4.0)  # Máximo de 4
    
    ig_total = parcela1 + parcela2
    
    # IG não pode ser negativo
    if ig_total < 0:
        ig_total = 0
    
    # Arredondar para o inteiro mais próximo
    return round(ig_total)


def _obter_descricao_hrb(grupo: str, subgrupo: Optional[str]) -> str:
    """
    Retorna a descrição do material baseado no grupo HRB.
    """
    descricoes = {
        "A-1": "Misturas bem graduadas de fragmentos de pedra, pedregulho e areia",
        "A-2": "Materiais granulares com teor de silte e argila",
        "A-3": "Areia fina de praia ou dunas, sem finos plásticos",
        "A-4": "Solos siltosos pouco ou nada plásticos",
        "A-5": "Solos siltosos com alta compressibilidade",
        "A-6": "Solos argilosos plásticos",
        "A-7": "Solos argilosos altamente plásticos"
    }
    
    subgrupo_desc = {
        "a": " (predominantemente pedregulho)",
        "b": " (predominantemente areia)",
        "4": " (características siltosas)",
        "5": " (características siltosas, alta compressibilidade)",
        "6": " (características argilosas)",
        "7": " (características argilosas, alta plasticidade)"
    }
    
    desc = descricoes.get(grupo, "Material não classificado")
    
    if subgrupo:
        desc += subgrupo_desc.get(subgrupo, "")
    
    return desc


def _obter_avaliacao_subleito(grupo: str) -> str:
    """
    Retorna a avaliação da qualidade do material como subleito de pavimento.
    """
    avaliacoes = {
        "A-1": "Excelente a Bom",
        "A-2": "Bom a Regular",
        "A-3": "Regular",
        "A-4": "Regular a Pobre",
        "A-5": "Pobre",
        "A-6": "Regular a Pobre",
        "A-7": "Pobre a Muito Pobre"
    }
    
    return avaliacoes.get(grupo, "Não avaliado")

File: frontend/src/lib/calculations/classificacao-hrb.ts
/**
 * Módulo para Classificação de Solos segundo o Sistema HRB/AASHTO
 * Utilizado para classificação de solos para pavimentação
 * 
 * Referências:
 * - AASHTO M 145
 */

import type { ClassificacaoHRBInput, ClassificacaoHRBOutput } from '../schemas';

const EPSILON = 1e-9;

export function classificarHRB(dados: ClassificacaoHRBInput): ClassificacaoHRBOutput {
  try {
    const p200 = dados.pass_peneira_200;
    const p40 = dados.pass_peneira_40;
    const p10 = dados.pass_peneira_10;
    // Tratar campos vazios de LL/LP como 0 automaticamente
    // para evitar classificação imprecisa (A-2 sem subgrupo) quando o usuário
    // esquece de digitar "0" em um solo não plástico
    const ll = dados.ll ?? 0;
    const ip = dados.ip ?? 0;

    // Validações
    if (p200 < 0 || p200 > 100) {
      throw new Error('% passando na #200 deve estar entre 0 e 100.');
    }
    if (ll < 0) {
      throw new Error('LL não pode ser negativo.');
    }
    if (ip < 0) {
      throw new Error('IP não pode ser negativo.');
    }

    // Determina o grupo principal
    const [grupo, subgrupo] = determinarGrupoHRB(p200, p40, p10, ll, ip);

    // Calcula o Índice de Grupo (IG) após determinar o grupo
    const ig = calcularIndiceGrupo(grupo, subgrupo, p200, ll, ip);

    // Monta a classificação completa
    let classificacao = subgrupo ? `${grupo}-${subgrupo}` : grupo;

    // Adiciona o índice de grupo se aplicável
    if (ig > 0) {
      classificacao += ` (${ig})`;
    }

    // Obtém a descrição do grupo
    const descricao = obterDescricaoHRB(grupo, subgrupo);

    // Obtém avaliação como subleito
    const avaliacao_subleito = obterAvaliacaoSubleito(grupo, subgrupo);

    return {
      classificacao,
      grupo_principal: grupo,
      subgrupo,
      indice_grupo: ig,
      descricao,
      avaliacao_subleito,
    };
  } catch (error) {
    return {
      erro: error instanceof Error ? error.message : 'Erro na classificação',
    };
  }
}

function determinarGrupoHRB(
  p200: number,
  p40: number | undefined,
  p10: number | undefined,
  ll: number,
  ip: number
): [string, string | undefined] {
  // MATERIAIS GRANULARES (≤ 35% passando na #200)
  if (p200 <= 35) {
    // 1. Tentar A-1-a
    if (
      (p10 === undefined || p10 <= 50) &&
      (p40 === undefined || p40 <= 30) &&
      p200 <= 15 &&
      ip <= 6
    ) {
      return ['A-1', 'a'];
    }

    // 2. Tentar A-1-b
    if (
      (p40 === undefined || p40 <= 50) &&
      p200 <= 25 &&
      ip <= 6
    ) {
      return ['A-1', 'b'];
    }

    // 3. Tentar A-3 (Conforme Nota 4, ele vem antes do A-2)
    if (
      (p40 !== undefined && p40 > 50) &&
      p200 <= 10 &&
      ip === 0
    ) {
      return ['A-3', undefined];
    }

    // 4. Se não for nenhum dos acima, DEVE ser A-2 (pois P200 <= 35)
    if (ll <= 40) {
      return ip <= 10 ? ['A-2', '4'] : ['A-2', '6'];
    } else {
      return ip <= 10 ? ['A-2', '5'] : ['A-2', '7'];
    }
  }

  // MATERIAIS SILTO-ARGILOSOS (> 35% passando na #200)
  // Leitura da esquerda para a direita
  else {
    // 1. Tentar A-4
    if (ll <= 40 && ip <= 10) {
      return ['A-4', undefined];
    }

    // 2. Tentar A-5
    if (ll > 40 && ip <= 10) {
      return ['A-5', undefined];
    }

    // 3. Tentar A-6
    if (ll <= 40 && ip > 10) {
      return ['A-6', undefined];
    }

    // 4. Deve ser A-7 (P200 > 35, LL > 40, IP > 10)
    // Nota 3: Subgrupos do A-7
    if (ip < ll - 30) {
      return ['A-7', '5'];
    } else {
      return ['A-7', '6'];
    }
  }
}

function calcularIndiceGrupo(
  grupo: string,
  subgrupo: string | undefined,
  p200: number,
  ll: number,
  ip: number
): number {
  // IG é sempre 0 para A-1-a, A-1-b e A-3
  if (grupo === 'A-1' || grupo === 'A-3') {
    return 0;
  }

  // Fórmula geral
  let ig = (p200 - 35) * (0.2 + 0.005 * (ll - 40)) + 0.01 * (p200 - 15) * (ip - 10);

  // Para A-2-4 e A-2-5, somente a parcela do IP é usada
  if (grupo === 'A-2' && (subgrupo === '4' || subgrupo === '5')) {
    ig = 0.01 * (p200 - 15) * (ip - 10);
  }

  // IG não pode ser negativo
  if (ig < 0) {
    return 0;
  }

  // Arredondar para o inteiro mais próximo
  return Math.round(ig);
}

function obterDescricaoHRB(grupo: string, subgrupo: string | undefined): string {
  // A-1
  if (grupo === 'A-1') {
    if (subgrupo === 'a') return 'Fragmentos de rocha ou pedregulho, com ou sem ligante fino bem graduado.';
    if (subgrupo === 'b') return 'Predominância de areia grossa, com ou sem ligante de solo bem graduado.';
    return 'Materiais granulares bem graduados.';
  }

  // A-3
  if (grupo === 'A-3') {
    return 'Areia fina sem finos siltosos/argilosos ou com uma quantidade muito pequena de finos.';
  }

  // A-2
  if (grupo === 'A-2') {
    if (subgrupo === '4') return 'Pedregulhos e areias com finos siltosos ou pouco plásticos (apresentam características dos grupos A-4 e A-5).';
    if (subgrupo === '5') return 'Pedregulhos e areias com finos siltosos ou pouco plásticos (apresentam características dos grupos A-4 e A-5).';
    if (subgrupo === '6') return 'Pedregulhos e areias com finos argilosos ou moderadamente plásticos (apresentam características dos grupos A-6 e A-7).';
    if (subgrupo === '7') return 'Pedregulhos e areias com finos argilosos ou moderadamente plásticos (apresentam características dos grupos A-6 e A-7).';
    return 'Pedregulhos e areias com conteúdo de siltes e argilas.';
  }

  // A-4
  if (grupo === 'A-4') {
    return 'Solo siltoso, moderadamente ou nada plástico. Geralmente com mais de 75% de finos.';
  }

  // A-5
  if (grupo === 'A-5') {
    return 'Solo siltoso, de caráter diatomáceo ou micáceo e pode ser altamente elástico ("borrachudo"), como indicado pelo alto LL.';
  }

  // A-6
  if (grupo === 'A-6') {
    return 'Solo argiloso plástico. Geralmente apresentam alta variação de volume entre os estados úmido e seco.';
  }

  // A-7
  if (grupo === 'A-7') {
    if (subgrupo === '5') return 'Solo argiloso elástico ("borrachudo"). Índice de plasticidade moderado, mas alta elasticidade.';
    if (subgrupo === '6') return 'Solo argiloso muito plástico. Sujeitos à altíssima variação de volume.';
    return 'Solo argiloso plástico ou elástico.';
  }

  return 'Material não classificado';
}

function obterAvaliacaoSubleito(grupo: string, subgrupo: string | undefined): string {
  // Materiais Granulares (Excelente a Bom)
  // A-1-a, A-1-b, A-3
  if (grupo === 'A-1' || grupo === 'A-3') {
    return 'Excelente a Bom';
  }

  // A-2
  if (grupo === 'A-2') {
    // A-2-4 e A-2-5 são Excelente a Bom
    if (subgrupo === '4' || subgrupo === '5') {
      return 'Excelente a Bom';
    }
    // A-2-6 e A-2-7 e qualquer outro caso são Regular a Mau
    return 'Regular a Mau';
  }

  // Materiais Silto-Argilosos (Regular a Mau)
  // A-4, A-5, A-6, A-7
  return 'Regular a Mau';
}

